### ISSUE TITLE
Extend `val_check_interval` support

### ISSUE THREAD

**Issue by amoshyc** [PyTorchLightning/pytorch-lightning#8135]
## 🐛 Bug

When I train the model by specifying number of training step instead of epoch, `val_check_interval` behaves strangely. Please see the following colab:

https://colab.research.google.com/drive/1I0ySRH03T9LdXHoGwCp3Q242dHEHWP_0?usp=sharing

In the code, I log the `global_step` on each validation.

```python
class Model(pl.LightningModule):
    ...

    def on_validation_start(self):
        print('global_step:', self.global_step)
```

I set Trainer's `max_steps` to `100` and `val_check_interval` to `10`.
But when I run the cells In[4] and In[5], the outputs are different.
The only different between In[4] and In[5] is the number of samples of the dataset which should not be the reason.

In[4]:
```python
train_set = RandomDataset(1, 40)
valid_set = RandomDataset(1, 40)
train_loader = torch.utils.data.DataLoader(train_set, batch_size=2)
valid_loader = torch.utils.data.DataLoader(valid_set, batch_size=2)

trainer = pl.Trainer(
    gpus=1,
    max_steps=100,
    val_check_interval=10,
    num_sanity_val_steps=0,
    log_every_n_steps=10,
    progress_bar_refresh_rate=0,
)
model = Model()
trainer.fit(model, train_loader, valid_loader)
```

Out[4]:
```
global_step: 9
global_step: 19
global_step: 29
global_step: 39
global_step: 49
global_step: 59
global_step: 69
global_step: 79
global_step: 89
global_step: 99
```

In[5]:
```python
train_set = RandomDataset(1, 32)
valid_set = RandomDataset(1, 32)
train_loader = torch.utils.data.DataLoader(train_set, batch_size=2)
valid_loader = torch.utils.data.DataLoader(valid_set, batch_size=2)

trainer = pl.Trainer(
    gpus=1,
    max_steps=100,
    val_check_interval=10,
    num_sanity_val_steps=0,
    log_every_n_steps=10,
    progress_bar_refresh_rate=0,
)
model = Model()
trainer.fit(model, train_loader, valid_loader)
```

Out[5]:
```
global_step: 9
global_step: 25
global_step: 41
global_step: 57
global_step: 73
global_step: 89
```

### Expected behavior

Since I specify `max_steps` and set `val_check_interval` to a integer, I expect that result is the same as Out[4] no matter the number of samples in the dataset. The [doc](https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html#val-check-interval) says that `val_check_interval` specifies the number of **training step** between validations, so Out[5] should be the same as Out[4].

I also expect that the number of times validation performed should be the same. BTW, the x-axis in the tensorboard are also wrong. You can see that in Out[9].

### Environment

```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.9.0+cu102
	- pytorch-lightning: 1.3.7post0
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Sat Jun 5 09:50:34 PDT 2021
```

cc @borda @tchaton

**Comment by tchaton** [comment]
Dear @amoshyc,

Thanks for reporting this issue. Would be interested in creating a fix PR ? 
I believe this shouldn't be too hard and a great way to learn more about Lightning.

Best.
T.C

**Comment by MirMustafaAli** [comment]
Hi @tchaton , 
Can i work on this issue as it is labelled as _good first issue_. ?

**Comment by amoshyc** [comment]
Sorry for the late reply. I'm so busy doing my research. I haven't have time to fix it.
The last time I check the code, it seems that the bug is [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/batch/training_batch_loop.py#L441-L453) and someone already notice it. 
But I'm not quite sure. I'm not working on it.

**Comment by tchaton** [comment]
Dear @MirMustafaAli,

Yes, feel free to work on this one and ping us if you get blocked :)

Best,
T.C

**Comment by MirMustafaAli** [comment]
Thanks

**Comment by MirMustafaAli** [comment]
@tchaton  
The issue seems to be in method __should_check_fx_ where it _batch idx_ of current epoch is being passed as parameter  . i was able to resolve it by passing  _total_batch_idx_  of epoch training loop to the method. 

https://github.com/PyTorchLightning/pytorch-lightning/blob/6b47cbe3ca8aa3fd82211bc9fa32e734753a6950/pytorch_lightning/loops/epoch/training_epoch_loop.py#L169-L170

Feedback required on solution

**Comment by carmocca** [comment]
```python
train_set = RandomDataset(1, 32)
valid_set = RandomDataset(1, 32)
train_loader = torch.utils.data.DataLoader(train_set, batch_size=2)
valid_loader = torch.utils.data.DataLoader(valid_set, batch_size=2)
```

Given this definition, your dataset has `32` batches. With `batch_size=2`, that gives us `32/2=16` batches per epoch.

Since you set `val_check_interval=10`, and there are 16 batches, you only have one chance to run validation in the epoch (in the 10th batch, index 9).

The issue here is that the docs definition is ambiguous

```python
            val_check_interval: How often to check the validation set. Use float to check within a training epoch,
                use int to check every n steps (batches).
```

"use int to check every n steps (batches)" does not explicitly state whether this is within an epoch or between epochs. The current implementation is for within an epoch.

Changing this would be a breaking change, and I can image users wanting to be able to choose which one to do.

cc @PyTorchLightning/core-contributors

**Comment by awaelchli** [comment]
@carmocca note there is also the `check_val_every_n_epochs` which is currently not mutually exclusive with val_check_interval if I understand correctly. If we did exactly what the docs said, we would have to make them mutually exclusive somehow, right?

**Comment by carmocca** [comment]
> we would have to make them mutually exclusive somehow, right?

I think so.

These are all the current options

- check every n batches across epochs: Not implemented. What this issue is requesting. Potentially `val_check_interval=n, check_val_every_n_epoch=None`?
- check every n batches per epoch: `val_check_interval=n, check_val_every_n_epoch=1`
- check every n epochs: `val_check_interval=1.0, check_val_every_n_epoch=n`
- check every n% batches across epochs: Not possible. Potentially `val_check_interval=0.n, check_val_every_n_epoch=None`?
- check every n% batches per epoch: `val_check_interval=0.n, check_val_every_n_epoch=1`
- check every n% epochs: Not possible. Potentially `val_check_interval=1.0, check_val_every_n_epoch=0.n`?

**Comment by MirMustafaAli** [comment]
@carmocca  is it possible to have one option with more preference over another?

if `check_val_every_n_epoch` has been provided it automatically supersedes `val_check_interval`.  making it generic and simple for user to use and understand. 

These are few things we might have to change. Should we implement it 
- We will have to update docs stating the explicitly about it.

**Comment by tchaton** [comment]
Hey @MirMustafaAli,

I believe the simplest to implement to resolve this issue is:

```
check every n batches across epochs: Not possible. Potentially val_check_interval=n, check_val_every_n_epoch=None?
```

**Comment by MirMustafaAli** [comment]
@tchaton 
will we explicitly explain it in docs?

**Comment by carmocca** [comment]
> will we explicitly explain it in docs?

Yes, we should describe all the possible options and the flags to set to achieve each of them

**Comment by MirMustafaAli** [comment]
> Hey @MirMustafaAli,
> 
> I believe the simplest to implement to resolve this issue is as shared by @carmocca.
> 
> ```
> check every n batches across epochs: Not possible. Potentially val_check_interval=n, check_val_every_n_epoch=None?
> ```


Shall i start with this options as a start?

**Comment by csvance** [comment]
So if I understand correctly, right now there is no good way to handle training with max_steps & val_check_interval where max_steps > len(dataloader)? 

Ideally we want to always trigger validation after the final step. So for instance, if I train 30,000 steps, validate every 3000 steps, and the length of the dataloader is 20,000, validation will happen at the following steps:

- 3000
- 6000
- 9000
- 12000
- 15000
- 18000 
- 23000 
- 26000
- 29000

Right now for this workflow, I manually run validation after trainer.fit finishes and save an additional checkpoint. 

In terms of potential workarounds, I could have my dataset lie about its length and report some rediculously high number, then take the modulo of the requested index with the true size of the dataset in order to make it valid. Then all steps would fit inside a single "epoch" and things should just work.

**Comment by mitchelldehaven** [comment]
What's the status on this? Like @csvance pointed out, whenever a new epoch starts it throws of doing validation by update step count.

**Comment by carmocca** [comment]
It's labeled as "help wanted", meaning, waiting for somebody to start working on it.

**Comment by mitchelldehaven** [comment]
Is it fine if I start working on it then?

**Comment by carmocca** [comment]
Yes. You can take one of the not possible/implemented options I described here:
https://github.com/PyTorchLightning/pytorch-lightning/issues/8135#issuecomment-888413064
and work on it.

**Comment by mitchelldehaven** [comment]
Okay, I think I will work on `check every n batches across epochs: Not implemented. What this issue is requesting. Potentially val_check_interval=n, check_val_every_n_epoch=None`, since that is the use case I am interested in.
